### ISSUE TITLE
[BUG] (RunAnd)WaitFor(Request/Response/RequestFinished)Async methods ignore exceptions thrown by action

### ISSUE THREAD

**Issue by maw136** [microsoft/playwright-dotnet#2530]
**Context:**
- Playwright Version: 1.31.0
- Operating System: any (Windows, Linux, MacOS)
- .NET version: .NET 7
- Browser: Edge (Chromium)

**Code Snippet**

```cs
using System.Threading.Tasks;
using Microsoft.Playwright;

class Program
{
    public static async Task Main()
    {
        using var playwright = await Playwright.CreateAsync();
        await using var browser = await playwright.Chromium.LaunchAsync();
        var page = await browser.NewPageAsync();
        var options = new() { Predicate = r => true };
        var wait = page.RunAndWaitForRequestFinishedAsync(()=>throw new Exception(), options)
        await wait;
        // dead waiting - only timeout and no original exception
    }
}
```

**Description**

`(RunAnd)WaitFor(Request/Response/RequestFinished)Async` method family are implemented by `InnerWaitForEventAsync` in such a way that ignores exceptions thrown by action. If the action ends up in Failed state (the resulting `Task` has failed because of some `Exception`) the actual Wait inside Playwright doesn't inspect the exception and the failed path ends up in timeout state. 
The Waiter makes that happen by throwing `TimeoutException`. 

The actual BUG is the fact that the Waiter still waits for some request to filter - but the action that was supposed to trigger some chit-chat that would caused some Requests has thrown in this scenario.

Action throwing -> Exception captured and set on a transient task in WhenAll -> the Waiter waits until timeout.

Secondly there is no `CancellationToken` to cancel Wait when a caller of `(...)WaitFor(...)` has some knowledge that the wait became pointless. There is some meaningful amount of time that can be recovered by providing a cancellation here.

***So, to summarize:***
1. Exceptions from action are missing
2. No cancellation option for Waits

***The exact places in code:***
playwright-dotnet\src\Playwright\Core\Page.cs : 435-438
playwright-dotnet\src\Playwright\Core\Waiter.cs - whole implementation - there should be a `RejectOnCancellation` or something along those lines.

Patched by: https://github.com/microsoft/playwright-dotnet/pull/2535

**Comment by maw136** [comment]
I would like to. I am currently looking into the test framework. I am considering a TDD path for patch development.

**Comment by maw136** [comment]
I'm adding this comment for searchability purpose, methods mentioned:
(RunAnd)WaitFor(Request/Response/RequestFinished)Async
(RunAnd)WaitFor(Request/Response/RequestFinished)Async
(RunAnd)WaitFor(Request/Response/RequestFinished)Async

**Comment by maw136** [comment]
I have attached a preliminary solution for the issue.
Moreover, I have added necessary unit tests and extended some existing tests as well.

**Comment by maw136** [comment]
I have attached a preliminary solution for the issue.
Moreover I have added necessary unit tests and extended some existing tests as well.

**Comment by maw136** [comment]
https://github.com/microsoft/playwright-dotnet/pull/2535 depends on :
https://github.com/microsoft/playwright-dotnet/pull/2538
https://github.com/microsoft/playwright-dotnet/pull/2540
https://github.com/microsoft/playwright-dotnet/pull/2542

**Comment by maw136** [comment]
PR approved, awaiting merge and release.

**Comment by mxschmitt** [comment]
Thanks!
